api/processors/actions/teacher/finished_but_stays_in_project.py

```diff
@@ -0,0 +1,33 @@
+from django.db import transaction
+
+from api.models import Teacher
+from api.models.choices.log_event_type import TeacherLogEventType
+from api.models.choices.status.project import TeacherProjectStatus
+from api.models.log_event import TeacherLogEvent
+from api.processors.actions.teacher import TeacherActionProcessor
+
+
+class TeacherFinishedStudiesButStaysInProjectProcessor(TeacherActionProcessor):
+    def __init__(self, teacher: Teacher):
+        super().__init__(teacher)
+
+    @transaction.atomic
+    def process(self) -> None:
+        self._update_groups()
+        self._create_log_events()
+        self._set_statuses()
+
+    def _update_groups(self) -> None:
+        if self.teacher.has_groups:
```

Comment:
This will never be executed since we check it in the view but I still implemented it in case we will use it separately. Another option is to throw an exception here

Reply:
I would throw an exception. If there are groups here, it is a coding error here in django_webapps, not user error (user = dashboard developer that programmed the request). As for which exception, I'm not sure. Maybe `ValueError`